apis/placement/v1alpha1/stagerollout_types.go

```diff
@@ -0,0 +1,135 @@
+// to target clusters.
+type StageRollout struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The desired state of StageRollout.
+	// +required
+	Spec StageRolloutSpec `json:"spec"`
+
+	// The observed status of StageRollout.
+	// +required
+	Status StageRolloutStatus `json:"status"`
+}
+
+// StageRolloutSpec defines the desired the rollout sequence stage by stage.
+type StageRolloutSpec struct {
+	// Stage rollout configurations for each rollout stage.
+	// +required
+	StageRollout []StageRolloutConfig `json:"stages"`
+
+	// The wait time between each stage is completed.
+	// +optional
+	WaitTimeBetweenStage *metav1.Duration `json:"waitTimeBetweenStage,omitempty"`
+
+	// TODO: Add alerting configuration.
+
+	// TODO: Add health-check configuration.
+}
+
+// StageRolloutConfig describes a single rollout stage group configuration.
+type StageRolloutConfig struct {
+	// The name of the stage. This should be unique within the StageRollout.
+	// +required
+	Name string `json:"name"`
+
+	// LabelSelector is a label query over all the joined member clusters. Clusters matching the query are selected.
+	// We don't check if there are overlap between clusters selected by different stages.
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+
+	// The label key used to sort the selected clusters.
+	// The clusters within the stage are updated following the rule below:
+	//   - primary: Ascending order based on the value of the label key, interpreted as integers if present.
+	//   - secondary: Ascending order based on the name of the cluster if the label key is absent.
+	// +optional
+	SortingLabelKey *string `json:"sortingLabelKey,omitempty"`
+
+	// The maximum number of clusters that can be updated simultaneously within this stage
+	// default is 1.
+	// +optional
+	MaxParallel *intstr.IntOrString `json:"maxParallel,omitempty"`
+
+	// The wait time after all the clusters in this stage are updated.
+	// This will override the global waitTimeBetweenStage for this specific stage
+	// +optional
+	WaitTime *metav1.Duration `json:"waitTime,omitempty"`
+}
+
+// StageRolloutStatus defines the observed state of the StageRollout.
+type StageRolloutStatus struct {
+	// The stage that is in the middle of the rollout.
+	CurrentStage int `json:"stage"`
+
+	// the clusters that are in the current stage.
+	ClustersInCurrentStage []string `json:"clustersInCurrentStage"`
+
+	// the clusters that are actively updating.
+	// +optional
+	CurrentUpdatingClusters []string `json:"currentUpdatingClusters,omitempty"`
+
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	//
+	// Conditions is an array of current observed conditions for ClusterResourceBinding.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions"`
+}
+
+// StageRolloutConditionType identifies a specific condition of the StageRollout.
+type StageRolloutConditionType string
+
+const (
```

Comment:
Hi Ryan! Just for clarification, from the comments it seems that `StageRolloutWaiting` is just a special case of `StageRollingOut` being true?